### ISSUE TITLE
Support for custom types in properties

### ISSUE THREAD

**Issue by redxdev** [bjorn/tiled#2199]
One of the problems I've personally run into with the property editor (and object types, by extension) is that the available types for each property (string, int, float, bool, etc) don't always cover all the different types of things I'd like to store into properties. There are plenty of existing issues about adding more types (arrays, object references, etc) but inevitably there will always be one more type to represent in properties (for example: vector or matrix values)

In my projects I tend to have a very generic tiled-to-custom-map-format parser that doesn't know exactly what type each property is supposed to be, so I end up writing a parser for the "string" type that supports putting vectors or other types into string properties (so something like `(0, 1)` would parse out to a vector). The problem with this is that I also have to support normal string properties, so that involves also writing out a way to escape parenthesis (for example) just in case I need to start an actual string with a `(`.

I'm proposing adding support for "custom" types that map to any existing type of property but have a different type name. In my mind this would work by having the `type` attribute on properties be set to `custom` and have an additional attribute named `customType` set to whatever the user specified. This would massively simplify writing custom property types.

**Comment by bjorn** [comment]
Yeah, I totally agree this can be useful as well, in addition to supporting custom compound types (#489), arrays (#1493) and enums (#1211). All these things (apart from arrays, possibly) have in common that we need a place where new property types can be defined. It would be nice if we can add this as part of the support for projects (#1665), which is planned to be the main feature for Tiled 1.4.

**Comment by bjorn** [comment]
Due to the addition of a number of other features and the changes having taken longer than estimated, I'm going to take this out of the roadmap for 1.4. I'd prefer to release Tiled 1.4 soon and will aim to have this in Tiled 1.5 instead.

**Comment by bjorn** [comment]
I've just merged some major work towards this feature as part of #3215, which for now only adds support for custom enum and classes, but can be extended to also support defining custom types based on existing types.
